src/namespaced_hash.rs

```diff
@@ -65,203 +64,273 @@ impl MerkleHash for NamespacedSha2Hasher {
-        let mut out = Self([0u8; NAMESPACED_HASH_LEN]);
-        out.0[0..NAMESPACE_ID_LEN].copy_from_slice(min_namespace.as_ref());
-        out.0[NAMESPACE_ID_LEN..2 * NAMESPACE_ID_LEN].copy_from_slice(max_namespace.as_ref());
-        out
+impl<const NS_ID_SIZE: usize> NamespacedHash<NS_ID_SIZE> {
+    pub const fn size() -> usize {
+        2 * NS_ID_SIZE + HASH_LEN
     }
-    pub fn min_namespace(&self) -> NamespaceId {
-        let mut out = [0u8; NAMESPACE_ID_LEN];
-        out.copy_from_slice(&self.0[..NAMESPACE_ID_LEN]);
-        NamespaceId(out)
+
+    pub const fn empty() -> Self {
+        Self {
+            min_ns: NamespaceId([0; NS_ID_SIZE]),
+            max_ns: NamespaceId([0; NS_ID_SIZE]),
+            hash: [
+                227, 176, 196, 66, 152, 252, 28, 20, 154, 251, 244, 200, 153, 111, 185, 36, 39,
+                174, 65, 228, 100, 155, 147, 76, 164, 149, 153, 27, 120, 82, 184, 85,
+            ],
+        }
     }
 
-    pub fn max_namespace(&self) -> NamespaceId {
-        let mut out = [0u8; NAMESPACE_ID_LEN];
-        out.copy_from_slice(&self.0[NAMESPACE_ID_LEN..2 * NAMESPACE_ID_LEN]);
-        NamespaceId(out)
+    pub fn with_min_and_max_ns(
+        min_ns: NamespaceId<NS_ID_SIZE>,
+        max_ns: NamespaceId<NS_ID_SIZE>,
+    ) -> Self {
+        Self {
+            min_ns,
+            max_ns,
+            ..Default::default()
+        }
     }
 
-    fn set_hash(&mut self, hash: &[u8]) {
-        self.0[2 * NAMESPACE_ID_LEN..].copy_from_slice(hash)
+    pub fn min_namespace(&self) -> NamespaceId<NS_ID_SIZE> {
+        self.min_ns
     }
 
-    pub fn empty() -> Self {
-        EMPTY_ROOT.clone()
+    pub fn max_namespace(&self) -> NamespaceId<NS_ID_SIZE> {
+        self.max_ns
     }
 
-    pub fn contains(&self, namespace: NamespaceId) -> bool {
+    fn set_hash(&mut self, new_hash: &[u8]) {
+        self.hash.copy_from_slice(new_hash)
+    }
+
+    pub fn contains(&self, namespace: NamespaceId<NS_ID_SIZE>) -> bool {
         self.min_namespace() <= namespace
             && self.max_namespace() >= namespace
             && !self.is_empty_root()
     }
 
     pub fn is_empty_root(&self) -> bool {
-        self == &EMPTY_ROOT
+        self == &Self::empty()
     }
 
-    pub fn hash_leaf(raw_data: impl AsRef<[u8]>, namespace: NamespaceId) -> Self {
+    pub fn hash_leaf(raw_data: impl AsRef<[u8]>, namespace: NamespaceId<NS_ID_SIZE>) -> Self {
         let mut output = NamespacedHash::with_min_and_max_ns(namespace, namespace);
         let mut hasher = Hasher::new_with_prefix(LEAF_DOMAIN_SEPARATOR);
-        hasher.update(namespace.as_ref());
+        hasher.update(namespace.as_ref()); // TODO: shouldn't this be hashed 2 times?
         hasher.update(raw_data.as_ref());
         output.set_hash(hasher.finalize().as_ref());
         output
     }
-}
 
-impl AsRef<[u8]> for NamespacedHash {
-    fn as_ref(&self) -> &[u8] {
-        self.0.as_ref()
+    pub fn iter(&self) -> impl Iterator<Item = &u8> {
```

Comment:
is this iterator for the internal bytes? i got confused when i saw it. i think we can have a `to_vec` instead

Reply:
Sure

Reply:
I think I prefer `to_bytes` as a name

Reply:
Ah, this isn't really working, because it would have to be `Self::size()` but putting const functions in place of generic parameters is not supported

Reply:
went with `to_vec`, hopefully malloc won't hurt us